dist_ir/backend/torch.py

```diff
@@ -397,34 +451,77 @@ def run_function(
         value_map[v] = x
     assert len(fn.inputs) == len(inputs)
 
-    def print_memory_usage():
-        t = torch.cuda.get_device_properties(0).total_memory
-        r = torch.cuda.memory_reserved(0)
-        a = torch.cuda.memory_allocated(0)
-        print(f"Total: {t} Reserved: {r} Allocated: {a} Free: {r-a}")
+    if ctx.measure_peak_memory:
+        torch.cuda.synchronize(device=rank)
+        memory_usage = get_memory_usage(rank)
+        allocated_memory = memory_usage.allocated
+        peak_memory = allocated_memory
+    else:
+        peak_memory = 0.0
+
+    recv_output_values = set()
+    op_runtime_events = []
 
     # Run ops
     for op in fn.ops:
         inputs = tuple(value_map[v] for v in op.inputs)
         kwargs = {} if op.attributes is None else {**op.attributes}
         kwargs["ctx"] = ctx
 
+        if record_op_runtimes:
+            add_event(ctx, op_runtime_events)
+        if op.op_type == "RecvP2P":
+            if op in recv_buffers:
+                buffer = recv_buffers[op]
+            else:
+                buffer = _allocate_recv_buffer(kwargs["shape"], kwargs["dtype"], ctx)
+                if ctx.use_gpu:
+                    torch.cuda.synchronize(device=rank)
```

Comment:
I think the point here is to make sure the buffer is fully allocated before passing it as an argument to the recv op - maybe this isn't necessary but in any case it only happens once for the first iteration

Reply:
Yeah, I'd be surprised if this was necessary. Even if the allocation hasn't happened yet by the time `_recv` is called, the `dist.broadcast` kernel will only execute after the allocation since we're using a single GPU stream.

Reply:
That's true, I'll remove this then